File: codemem/viewer_routes/memory.py
from __future__ import annotations

from typing import Any, Protocol
from urllib.parse import parse_qs

from ..config import load_config
from ..db import from_json
from ..store import MemoryStore


class _ViewerHandler(Protocol):
    def _send_json(self, payload: dict[str, Any], status: int = 200) -> None: ...


def _attach_session_fields(store: MemoryStore, items: list[dict[str, Any]]) -> None:
    session_ids: list[int] = []
    seen: set[int] = set()
    for item in items:
        value = item.get("session_id")
        if value is None:
            continue
        try:
            sid = int(value)
        except (TypeError, ValueError):
            continue
        if sid in seen:
            continue
        seen.add(sid)
        session_ids.append(sid)

    if not session_ids:
        return

    placeholders = ",".join("?" for _ in session_ids)
    rows = store.conn.execute(
        f"SELECT id, project, cwd FROM sessions WHERE id IN ({placeholders})",
        session_ids,
    ).fetchall()
    by_session: dict[int, dict[str, str]] = {}
    for row in rows:
        try:
            sid = int(row["id"])
        except (TypeError, ValueError):
            continue
        project_raw = row["project"] or ""
        project = store._project_basename(project_raw.strip()) if project_raw else ""
        cwd = row["cwd"] or ""
        by_session[sid] = {"project": project, "cwd": cwd}

    for item in items:
        value = item.get("session_id")
        if value is None:
            continue
        try:
            sid = int(value)
        except (TypeError, ValueError):
            continue
        fields = by_session.get(sid)
        if not fields:
            continue
        item.setdefault("project", fields.get("project") or "")
        item.setdefault("cwd", fields.get("cwd") or "")


def _attach_ownership_fields(store: MemoryStore, items: list[dict[str, Any]]) -> None:
    for item in items:
        item["owned_by_self"] = store.memory_owned_by_self(item)


def _apply_scope_filter(
    store: MemoryStore, filters: dict[str, Any] | None, scope: str | None
) -> tuple[dict[str, Any], bool]:
    normalized = str(scope or "all").strip().lower()
    if normalized not in {"all", "mine", "theirs", "shared"}:
        return {}, False
    scoped = dict(filters or {})
    if normalized == "mine":
        scoped["ownership_scope"] = "mine"
    elif normalized == "theirs":
        scoped["ownership_scope"] = "theirs"
    elif normalized == "shared":
        scoped["include_visibility"] = ["shared"]
    return scoped, True


def handle_get(handler: _ViewerHandler, store: MemoryStore, path: str, query: str) -> bool:
    # Compatibility endpoints used by the bundled web UI.
    if path == "/api/memories":
        path = "/api/observations"

    if path == "/api/sessions":
        params = parse_qs(query)
        limit = int(params.get("limit", ["20"])[0])
        sessions = store.all_sessions()[:limit]
        for item in sessions:
            item["metadata_json"] = from_json(item.get("metadata_json"))
        handler._send_json({"items": sessions})
        return True

    if path == "/api/projects":
        sessions = store.all_sessions()
        projects = sorted(
            {
                store._project_basename(p.strip())
                for s in sessions
                if (p := s.get("project"))
                and isinstance(p, str)
                and p.strip()
                and not p.strip().lower().startswith("fatal:")
                and store._project_basename(p.strip())
            }
        )
        handler._send_json({"projects": projects})
        return True

    if path == "/api/observations":
        params = parse_qs(query)
        try:
            limit = max(1, int(params.get("limit", ["20"])[0]))
            offset = max(0, int(params.get("offset", ["0"])[0]))
        except ValueError:
            handler._send_json({"error": "limit and offset must be int"}, status=400)
            return True
        project = params.get("project", [None])[0]
        scope = params.get("scope", ["all"])[0]
        kinds = [
            "bugfix",
            "change",
            "decision",
            "discovery",
            "exploration",
            "feature",
            "refactor",
        ]
        obs_filters, valid_scope = _apply_scope_filter(
            store, {"project": project} if project else None, scope
        )
        if not valid_scope:
            handler._send_json({"error": "invalid_scope"}, status=400)
            return True
        items = store.recent_by_kinds(
            limit=limit + 1, kinds=kinds, filters=obs_filters, offset=offset
        )
        has_more = len(items) > limit
        if has_more:
            items = items[:limit]
        _attach_session_fields(store, items)
        _attach_ownership_fields(store, items)
        handler._send_json(
            {
                "items": items,
                "pagination": {
                    "limit": limit,
                    "offset": offset,
                    "next_offset": offset + len(items) if has_more else None,
                    "has_more": has_more,
                },
            }
        )
        return True

    if path == "/api/summaries":
        params = parse_qs(query)
        try:
            limit = max(1, int(params.get("limit", ["50"])[0]))
            offset = max(0, int(params.get("offset", ["0"])[0]))
        except ValueError:
            handler._send_json({"error": "limit and offset must be int"}, status=400)
            return True
        project = params.get("project", [None])[0]
        scope = params.get("scope", ["all"])[0]
        filters: dict[str, Any] = {"kind": "session_summary"}
        if project:
            filters["project"] = project
        scoped_filters, valid_scope = _apply_scope_filter(store, filters, scope)
        if not valid_scope:
            handler._send_json({"error": "invalid_scope"}, status=400)
            return True
        items = store.recent(limit=limit + 1, filters=scoped_filters, offset=offset)
        has_more = len(items) > limit
        if has_more:
            items = items[:limit]
        _attach_session_fields(store, items)
        _attach_ownership_fields(store, items)
        handler._send_json(
            {
                "items": items,
                "pagination": {
                    "limit": limit,
                    "offset": offset,
                    "next_offset": offset + len(items) if has_more else None,
                    "has_more": has_more,
                },
            }
        )
        return True

    if path == "/api/session":
        params = parse_qs(query)
        project = params.get("project", [None])[0]

        if project:
            prompts = store.conn.execute(
                "SELECT COUNT(*) AS total FROM user_prompts WHERE project = ?",
                (project,),
            ).fetchone()["total"]
            artifacts = store.conn.execute(
                """
                SELECT COUNT(*) AS total
                FROM artifacts
                JOIN sessions ON sessions.id = artifacts.session_id
                WHERE sessions.project = ?
                """,
                (project,),
            ).fetchone()["total"]
            memories = store.conn.execute(
                """
                SELECT COUNT(*) AS total
                FROM memory_items
                JOIN sessions ON sessions.id = memory_items.session_id
                WHERE sessions.project = ?
                """,
                (project,),
            ).fetchone()["total"]
            observations = store.conn.execute(
                """
                SELECT COUNT(*) AS total
                FROM memory_items
                JOIN sessions ON sessions.id = memory_items.session_id
                WHERE kind != 'session_summary'
                  AND sessions.project = ?
                """,
                (project,),
            ).fetchone()["total"]
        else:
            prompts = store.conn.execute("SELECT COUNT(*) AS total FROM user_prompts").fetchone()[
                "total"
            ]
            artifacts = store.conn.execute("SELECT COUNT(*) AS total FROM artifacts").fetchone()[
                "total"
            ]
            memories = store.conn.execute("SELECT COUNT(*) AS total FROM memory_items").fetchone()[
                "total"
            ]
            observations = store.conn.execute(
                "SELECT COUNT(*) AS total FROM memory_items WHERE kind != 'session_summary'"
            ).fetchone()["total"]
        total = int(prompts or 0) + int(artifacts or 0) + int(memories or 0)

        handler._send_json(
            {
                "total": total,
                "memories": int(memories or 0),
                "artifacts": int(artifacts or 0),
                "prompts": int(prompts or 0),
                "observations": int(observations or 0),
            }
        )
        return True

    if path == "/api/pack":
        params = parse_qs(query)
        context = params.get("context", [""])[0]
        if not context:
            handler._send_json({"error": "context required"}, status=400)
            return True
        config = load_config()
        try:
            limit = int(params.get("limit", [str(config.pack_observation_limit)])[0])
        except ValueError:
            handler._send_json({"error": "limit must be int"}, status=400)
            return True
        token_budget = params.get("token_budget", [None])[0]
        if token_budget in (None, ""):
            token_budget_value = None
        else:
            try:
                token_budget_value = int(token_budget)
            except ValueError:
                handler._send_json({"error": "token_budget must be int"}, status=400)
                return True
        project = params.get("project", [None])[0]
        scope = params.get("scope", ["all"])[0]
        pack_filters, valid_scope = _apply_scope_filter(
            store, {"project": project} if project else None, scope
        )
        if not valid_scope:
            handler._send_json({"error": "invalid_scope"}, status=400)
            return True
        pack = store.build_memory_pack(
            context=context,
            limit=limit,
            token_budget=token_budget_value,
            filters=pack_filters,
        )
        handler._send_json(pack)
        return True

    if path == "/api/memory":
        params = parse_qs(query)
        limit = int(params.get("limit", ["20"])[0])
        kind = params.get("kind", [None])[0]
        project = params.get("project", [None])[0]
        scope = params.get("scope", ["all"])[0]
        filters: dict[str, Any] = {}
        if kind:
            filters["kind"] = kind
        if project:
            filters["project"] = project
        scoped_filters, valid_scope = _apply_scope_filter(
            store, filters if filters else None, scope
        )
        if not valid_scope:
            handler._send_json({"error": "invalid_scope"}, status=400)
            return True
        items = store.recent(limit=limit, filters=scoped_filters)
        _attach_session_fields(store, items)
        _attach_ownership_fields(store, items)
        handler._send_json({"items": items})
        return True

    if path == "/api/artifacts":
        params = parse_qs(query)
        session_id = params.get("session_id", [None])[0]
        if not session_id:
            handler._send_json({"error": "session_id required"}, status=400)
            return True
        items = store.session_artifacts(int(session_id))
        handler._send_json({"items": items})
        return True

    return False


def handle_post(
    handler: _ViewerHandler, store: MemoryStore, path: str, payload: dict[str, Any] | None
) -> bool:
    if path != "/api/memories/visibility":
        return False
    if payload is None:
        handler._send_json({"error": "invalid json"}, status=400)
        return True
    memory_id = payload.get("memory_id")
    visibility = payload.get("visibility")
    try:
        resolved_memory_id = int(memory_id)
    except (TypeError, ValueError):
        handler._send_json({"error": "memory_id must be int"}, status=400)
        return True
    if not isinstance(visibility, str) or visibility.strip() not in {"private", "shared"}:
        handler._send_json({"error": "visibility must be private or shared"}, status=400)
        return True
    try:
        item = store.update_memory_visibility(
            resolved_memory_id,
            visibility=visibility.strip(),
        )
    except ValueError as exc:
        handler._send_json({"error": str(exc)}, status=400)
        return True
    except LookupError:
        handler._send_json({"error": "memory not found"}, status=404)
        return True
    except PermissionError:
        handler._send_json({"error": "memory not owned by self"}, status=403)
        return True
    item["owned_by_self"] = store.memory_owned_by_self(item)
    handler._send_json({"item": item})
    return True

File: packages/viewer-server/src/routes/memory.ts
/**
 * Memory routes — port of codemem/viewer_routes/memory.py.
 *
 * Routes:
 *   GET  /api/memories     — list observations (alias for /api/observations)
 *   GET  /api/observations — paginated observations by kind
 *   GET  /api/summaries    — paginated session summaries
 *   GET  /api/memory       — recent memories with optional kind/project filter
 *   GET  /api/sessions     — recent sessions
 *   GET  /api/session      — aggregate counts for a project
 *   GET  /api/projects     — distinct project names
 *   GET  /api/artifacts    — artifacts for a session
 *   GET  /api/pack         — build memory pack
 *   POST /api/memories/visibility — update memory visibility
 */

import {
	fromJson,
	type MemoryFilters,
	type MemoryItemResponse,
	type MemoryStore,
} from "@codemem/core";
import { Hono } from "hono";
import type { ViewerVariables } from "../middleware.js";

const app = new Hono<{ Variables: ViewerVariables }>();

// ---------------------------------------------------------------------------
// Helpers — mirrors Python's _attach_session_fields / _attach_ownership_fields
// ---------------------------------------------------------------------------

interface MemoryWithSessionFields extends MemoryItemResponse {
	project?: string;
	cwd?: string;
	owned_by_self?: boolean;
}

/**
 * Attach session project/cwd to memory items.
 * Python source: codemem/viewer_routes/memory.py lines 15-62
 */
function attachSessionFields(store: MemoryStore, items: MemoryWithSessionFields[]): void {
	const sessionIds: number[] = [];
	const seen = new Set<number>();
	for (const item of items) {
		const sid = item.session_id;
		if (sid == null || seen.has(sid)) continue;
		seen.add(sid);
		sessionIds.push(sid);
	}
	if (sessionIds.length === 0) return;

	const placeholders = sessionIds.map(() => "?").join(", ");
	const rows = store.db
		.prepare(`SELECT id, project, cwd FROM sessions WHERE id IN (${placeholders})`)
		.all(...sessionIds) as Array<{ id: number; project: string | null; cwd: string | null }>;

	const bySession = new Map<number, { project: string; cwd: string }>();
	for (const row of rows) {
		const project = (row.project ?? "").trim();
		// Python uses store._project_basename — just take the last path segment
		const basename = project ? (project.split("/").pop() ?? project) : "";
		bySession.set(row.id, { project: basename, cwd: row.cwd ?? "" });
	}

	for (const item of items) {
		const fields = bySession.get(item.session_id);
		if (!fields) continue;
		item.project ??= fields.project;
		item.cwd ??= fields.cwd;
	}
}

/**
 * Attach ownership flags to memory items.
 * Python source: codemem/viewer_routes/memory.py lines 65-67
 */
function attachOwnershipFields(store: MemoryStore, items: MemoryWithSessionFields[]): void {
	for (const item of items) {
		// Mirrors Python's memory_owned_by_self: checks origin_device_id
		item.owned_by_self = !item.origin_device_id || item.origin_device_id === store.deviceId;
	}
}

/**
 * Apply scope filter to memory filters.
 * Python source: codemem/viewer_routes/memory.py lines 70-83
 */
function applyScopeFilter(
	store: MemoryStore,
	baseFilters: MemoryFilters | undefined,
	scope: string | null,
): { filters: MemoryFilters; valid: boolean } {
	const normalized = (scope ?? "all").trim().toLowerCase();
	if (!["all", "mine", "theirs", "shared"].includes(normalized)) {
		return { filters: {}, valid: false };
	}
	const filters: MemoryFilters = { ...(baseFilters ?? {}) };
	if (normalized === "mine") {
		// TODO: ownership_scope filter not yet in TS buildFilterClauses
		// Python source: codemem/viewer_routes/memory.py line 78
	} else if (normalized === "theirs") {
		// TODO: ownership_scope filter not yet in TS buildFilterClauses
	} else if (normalized === "shared") {
		filters.include_visibility = ["shared"];
	}
	void store; // used by ownership_scope once ported
	return { filters, valid: true };
}

// ---------------------------------------------------------------------------
// Observation kinds (memories minus session_summary)
// ---------------------------------------------------------------------------

const OBSERVATION_KINDS = [
	"bugfix",
	"change",
	"decision",
	"discovery",
	"exploration",
	"feature",
	"refactor",
];

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function safeInt(value: string | undefined, fallback: number): number {
	if (value == null) return fallback;
	const n = Number.parseInt(value, 10);
	return Number.isNaN(n) ? fallback : n;
}

function intParam(
	c: { req: { query: (k: string) => string | undefined } },
	key: string,
	fallback: number,
): number | null {
	const raw = c.req.query(key);
	if (raw == null) return fallback;
	const n = Number.parseInt(raw, 10);
	if (Number.isNaN(n)) return null; // signal parse error
	return n;
}

// ---------------------------------------------------------------------------
// GET routes
// ---------------------------------------------------------------------------

app.get("/api/memories", (c) => {
	// Compatibility endpoint — redirect logic to /api/observations
	const url = new URL(c.req.url);
	url.pathname = "/api/observations";
	return c.redirect(url.pathname + url.search, 307);
});

app.get("/api/observations", (c) => {
	const store = c.get("store") as MemoryStore;
	const limit = intParam(c, "limit", 20);
	const offset = intParam(c, "offset", 0);
	if (limit === null || offset === null) {
		return c.json({ error: "limit and offset must be int" }, 400);
	}
	const clampedLimit = Math.max(1, limit);
	const clampedOffset = Math.max(0, offset);

	const project = c.req.query("project") ?? null;
	const scope = c.req.query("scope") ?? "all";

	const baseFilters: MemoryFilters | undefined = project ? { project } : undefined;
	const { filters, valid } = applyScopeFilter(store, baseFilters, scope);
	if (!valid) return c.json({ error: "invalid_scope" }, 400);

	const items = store.recentByKinds(OBSERVATION_KINDS, clampedLimit + 1, filters, clampedOffset);
	const hasMore = items.length > clampedLimit;
	const page = hasMore ? items.slice(0, clampedLimit) : items;
	const enriched = page as MemoryWithSessionFields[];
	attachSessionFields(store, enriched);
	attachOwnershipFields(store, enriched);

	return c.json({
		items: enriched,
		pagination: {
			limit: clampedLimit,
			offset: clampedOffset,
			next_offset: hasMore ? clampedOffset + page.length : null,
			has_more: hasMore,
		},
	});
});

app.get("/api/summaries", (c) => {
	const store = c.get("store") as MemoryStore;
	const limit = intParam(c, "limit", 50);
	const offset = intParam(c, "offset", 0);
	if (limit === null || offset === null) {
		return c.json({ error: "limit and offset must be int" }, 400);
	}
	const clampedLimit = Math.max(1, limit);
	const clampedOffset = Math.max(0, offset);

	const project = c.req.query("project") ?? null;
	const scope = c.req.query("scope") ?? "all";

	const baseFilters: MemoryFilters = { kind: "session_summary" };
	if (project) baseFilters.project = project;
	const { filters, valid } = applyScopeFilter(store, baseFilters, scope);
	if (!valid) return c.json({ error: "invalid_scope" }, 400);

	const items = store.recent(clampedLimit + 1, filters, clampedOffset);
	const hasMore = items.length > clampedLimit;
	const page = hasMore ? items.slice(0, clampedLimit) : items;
	const enriched = page as MemoryWithSessionFields[];
	attachSessionFields(store, enriched);
	attachOwnershipFields(store, enriched);

	return c.json({
		items: enriched,
		pagination: {
			limit: clampedLimit,
			offset: clampedOffset,
			next_offset: hasMore ? clampedOffset + page.length : null,
			has_more: hasMore,
		},
	});
});

app.get("/api/memory", (c) => {
	const store = c.get("store") as MemoryStore;
	const limit = safeInt(c.req.query("limit"), 20);
	const kind = c.req.query("kind") ?? null;
	const project = c.req.query("project") ?? null;
	const scope = c.req.query("scope") ?? "all";

	const baseFilters: MemoryFilters = {};
	if (kind) baseFilters.kind = kind;
	if (project) baseFilters.project = project;
	const { filters, valid } = applyScopeFilter(
		store,
		Object.keys(baseFilters).length > 0 ? baseFilters : undefined,
		scope,
	);
	if (!valid) return c.json({ error: "invalid_scope" }, 400);

	const items = store.recent(limit, filters) as MemoryWithSessionFields[];
	attachSessionFields(store, items);
	attachOwnershipFields(store, items);

	return c.json({ items });
});

app.get("/api/sessions", (c) => {
	const store = c.get("store") as MemoryStore;
	const limit = safeInt(c.req.query("limit"), 20);

	// TODO: Port allSessions() to TS store
	// Python source: codemem/viewer_routes/memory.py lines 91-98
	const rows = store.db
		.prepare("SELECT * FROM sessions ORDER BY started_at DESC LIMIT ?")
		.all(limit) as Array<Record<string, unknown>>;

	const items = rows.map((row) => ({
		...row,
		metadata_json: fromJson(row.metadata_json as string | null),
	}));

	return c.json({ items });
});

app.get("/api/projects", (c) => {
	const store = c.get("store") as MemoryStore;

	// TODO: Port allSessions() / _project_basename() to TS store
	// Python source: codemem/viewer_routes/memory.py lines 100-114
	const rows = store.db
		.prepare("SELECT project FROM sessions WHERE project IS NOT NULL AND project != ''")
		.all() as Array<{ project: string }>;

	const projects = [
		...new Set(
			rows
				.map((r) => r.project.trim())
				.filter((p) => p && !p.toLowerCase().startsWith("fatal:"))
				.map((p) => p.split("/").pop() ?? p)
				.filter(Boolean),
		),
	].sort();

	return c.json({ projects });
});

app.get("/api/session", (c) => {
	const store = c.get("store") as MemoryStore;
	const project = c.req.query("project") ?? null;

	// Aggregate counts — mirrors Python codemem/viewer_routes/memory.py lines 198-258
	const count = (sql: string, params: unknown[] = []): number => {
		const row = store.db.prepare(sql).get(...params) as { total: number } | undefined;
		return row?.total ?? 0;
	};

	let prompts: number;
	let artifacts: number;
	let memories: number;
	let observations: number;

	if (project) {
		prompts = count("SELECT COUNT(*) AS total FROM user_prompts WHERE project = ?", [project]);
		artifacts = count(
			`SELECT COUNT(*) AS total FROM artifacts
			 JOIN sessions ON sessions.id = artifacts.session_id
			 WHERE sessions.project = ?`,
			[project],
		);
		memories = count(
			`SELECT COUNT(*) AS total FROM memory_items
			 JOIN sessions ON sessions.id = memory_items.session_id
			 WHERE sessions.project = ?`,
			[project],
		);
		observations = count(
			`SELECT COUNT(*) AS total FROM memory_items
			 JOIN sessions ON sessions.id = memory_items.session_id
			 WHERE kind != 'session_summary' AND sessions.project = ?`,
			[project],
		);
	} else {
		prompts = count("SELECT COUNT(*) AS total FROM user_prompts");
		artifacts = count("SELECT COUNT(*) AS total FROM artifacts");
		memories = count("SELECT COUNT(*) AS total FROM memory_items");
		observations = count(
			"SELECT COUNT(*) AS total FROM memory_items WHERE kind != 'session_summary'",
		);
	}

	return c.json({
		total: prompts + artifacts + memories,
		memories,
		artifacts,
		prompts,
		observations,
	});
});

app.get("/api/artifacts", (c) => {
	const store = c.get("store") as MemoryStore;
	const sessionId = c.req.query("session_id");
	if (!sessionId) {
		return c.json({ error: "session_id required" }, 400);
	}
	const sid = Number.parseInt(sessionId, 10);
	if (Number.isNaN(sid)) {
		return c.json({ error: "session_id must be int" }, 400);
	}

	// TODO: Port sessionArtifacts() to TS store
	// Python source: codemem/viewer_routes/memory.py lines 322-330
	const rows = store.db
		.prepare("SELECT * FROM artifacts WHERE session_id = ? ORDER BY created_at DESC")
		.all(sid) as Array<Record<string, unknown>>;

	return c.json({ items: rows });
});

app.get("/api/pack", (c) => {
	const store = c.get("store") as MemoryStore;
	const context = c.req.query("context") ?? "";
	if (!context) {
		return c.json({ error: "context required" }, 400);
	}

	const limitRaw = c.req.query("limit");
	const limit = limitRaw ? Number.parseInt(limitRaw, 10) : undefined;
	if (limitRaw && (limit == null || Number.isNaN(limit))) {
		return c.json({ error: "limit must be int" }, 400);
	}

	const tokenBudgetRaw = c.req.query("token_budget");
	let tokenBudget: number | null = null;
	if (tokenBudgetRaw && tokenBudgetRaw !== "") {
		tokenBudget = Number.parseInt(tokenBudgetRaw, 10);
		if (Number.isNaN(tokenBudget)) {
			return c.json({ error: "token_budget must be int" }, 400);
		}
	}

	const project = c.req.query("project") ?? null;
	const scope = c.req.query("scope") ?? "all";
	const baseFilters: MemoryFilters | undefined = project ? { project } : undefined;
	const { filters, valid } = applyScopeFilter(store, baseFilters, scope);
	if (!valid) return c.json({ error: "invalid_scope" }, 400);

	const pack = store.buildMemoryPack(context, limit, tokenBudget, filters);
	return c.json(pack);
});

// ---------------------------------------------------------------------------
// POST routes
// ---------------------------------------------------------------------------

app.post("/api/memories/visibility", async (c) => {
	const store = c.get("store") as MemoryStore;
	let payload: Record<string, unknown>;
	try {
		payload = await c.req.json();
	} catch {
		return c.json({ error: "invalid json" }, 400);
	}

	const memoryId = payload.memory_id;
	const visibility = payload.visibility;

	if (memoryId == null || typeof memoryId !== "number") {
		return c.json({ error: "memory_id must be int" }, 400);
	}
	if (typeof visibility !== "string" || !["private", "shared"].includes(visibility.trim())) {
		return c.json({ error: "visibility must be private or shared" }, 400);
	}

	try {
		const item = store.updateMemoryVisibility(memoryId, visibility.trim());
		const enriched = item as MemoryWithSessionFields;
		enriched.owned_by_self =
			!enriched.origin_device_id || enriched.origin_device_id === store.deviceId;
		return c.json({ item: enriched });
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		if (msg === "memory not found") {
			return c.json({ error: "memory not found" }, 404);
		}
		if (msg.includes("not owned")) {
			return c.json({ error: "memory not owned by self" }, 403);
		}
		return c.json({ error: msg }, 400);
	}
});

export default app;
